### DISCUSSION TITLE
Creating spy/mock for a member.roles.add

### DISCUSSION THREAD

**Discussion by NicholasJohn16** [discordjs/discord.js#8875] [Q&A]
I'm trying to mock [member.roles.add][1] in a test using `vitest`, but it's not working. I'm pretty new to testing and some help would be appreciated.

`interaction` is a mock that's generated like [this][2].

```js
it('adds officer role if user doesn\'t have  role', () => {
    const command = {'id': 'officer', 'name': 'officer', 'type': 1};
    const discord = new MockDiscord({ command });
    const interaction = discord.getInteraction()
    const targetMember = discord.getGuildMember();
    
    interaction.targetMember = targetMember;  
    const spy = vi.spyOn(interaction.targetMember.roles, 'add').mockImplementation(() => true);

    process.env.OFFICER_ROLE_ID = 1;
    officer.execute(interaction);
    expect(spy).toHaveBeenCalled();
});
```

The test fails because `spy` is never called and I get an error cause there wasn't a token for the `add` request. Any ideas how to get it to work? Thanks.

  [1]: https://discord.js.org/#/docs/discord.js/main/class/GuildMemberRoleManager?scrollTo=add
  [2]: https://github.com/markkop/corvo-astral/blob/master/tests/mockDiscord.ts#L349

**Comment by ckohen** [comment]
`<member>.roles` is dynamically calculated at runtime (its a getter and the manager is recreated every time .roles is called), that is why you are seeing this error.

I can't think of a good way to mock this without having an entire mock api running (which is too much work)

**Reply by NicholasJohn16** [reply]
I was thinking it was something like that. Would you mind showing me where in the code roles is dynamically calculated? Thank you very much for your reply. Been scratching my head at this all weekend.

**Reply by ckohen** [reply]
https://github.com/discordjs/discord.js/blob/6bcebd48671021b1ce6cfed60b01ebe541f9be1e/packages/discord.js/src/structures/GuildMember.js#L113-L120
